### ISSUE TITLE
Why does graph sometimes show a steep downturn on the right hand side of chart?

### ISSUE THREAD

**Issue by simon-anz** [miguelmota/cointop#157]
![Screen Shot 2021-09-02 at 11 41 18 am](https://user-images.githubusercontent.com/69175652/131768044-9d81205a-ea78-402e-9451-f012f0e60d18.png)
I think this is quite unlikely to represent the actual prices/value.

**Comment by simon-anz** [comment]
Seems to be caused by uneven number of data points in the chart data:

```
2021/09/02 12:10:33 PortfolioChart() Bitcoin = 74 items
2021/09/02 12:10:33 PortfolioChart() Bitcoin Cash = 73 items
2021/09/02 12:10:33 PortfolioChart() 74 ==> [52719.95682714844 53407.706991943356 53041.254965087894 52534.5023425293 52083.594048339844 51995.50595898437 52150.22023657226 52279.433401855465 ...
51647.00320727539 51166.04865039063 51212.208677490235 51109.865504882815 50914.87169775391 51103.57940405273 51096.73617724609 50873.03044482422 50923.954295410156 51021.63647680664 47230.625]
```

**Comment by simon-anz** [comment]
Even deleting the cache and restarting, had the same problem (my default view is 3D portfolio)

**Comment by simon-anz** [comment]
Problem seems to be originating at data-source (but happens often enough that we should defend against it). Yesterday I was getting 74/73 items

```
ct.api.GetCoinGraphData(USD, BTC, Bitcoin, 1630305502, 1630564702) ===> 74
ct.api.GetCoinGraphData(USD, BCH, Bitcoin Cash, 1630305502, 1630564702) ===> 73
```

today

```
ct.api.GetCoinGraphData(USD, BTC, Bitcoin, 1630357554, 1630616754) ===> 74
ct.api.GetCoinGraphData(USD, BCH, Bitcoin Cash, 1630357554, 1630616754) ===> 74
```

Maybe we need to consider the x-axis (time) values, not just the y-axis (Price).

**Comment by simon-anz** [comment]
Even passing yesterdays start/end returns 74 items today.

**Comment by simon-anz** [comment]
The problem is the data returned (from coingecko at least) is not super-consistent about how the time values work (especially at the start end end):

```
GetCoinGraphData(USD, BTC, Bitcoin, 2021-08-30T16:38:22+10:00, 2021-09-02T16:38:22+10:00)
    2021-08-31T08:00:29+10:00 : 48446.238281
    2021-08-31T09:03:51+10:00 : 47673.679688
    2021-08-31T09:03:51+10:00 : 47673.679688
    2021-08-31T10:00:38+10:00 : 47124.253906
    2021-08-31T11:01:48+10:00 : 46968.371094
...
    2021-09-03T06:01:21+10:00 : 49333.996094
    2021-09-03T07:02:31+10:00 : 49549.394531
    2021-09-03T07:22:11+10:00 : 49556.132812
```

**Comment by simon-anz** [comment]
TLDR the data source is a bit inconsistent - the data should probably be resampled with a consistent time-interval.   I've got some code here that works OK, lifted from https://github.com/devtoolkits/downsample but would prefer to find a library for this. Thoughts?

**Comment by blockjoe** [comment]
I've noticed the odd steep downturn doesn't have to happen at the very end of a portfolio, but it can also happen pretty far into the chart.

This is just a simple example of two assets on the 6M chart, 1 BTC and 1000 GTC. I picked these since GTC launched on 5/24. In theory the peak at the very start of this chart shouldn't be possible since GTC wasn't around for this. The steep that happens 2/3 of the way through this graph lines up with about the amount of time that GTC has been available. 

On the webview of coingecko, when I select 180D, it only gives me 5/24 - 9/2. So I can imagine if you are asking for a time frame of data you're assuming that same time frame coming back, but that's probably not the case. This probably means having to check that data existed X months back, and if not zero padding the prices at the start for when the coin didn't exist. 

 A cursory for time series libraries for go didn't give me much luck. [traces](https://github.com/datascopeanalytics/traces) is a Python package designed to handle this problem (almost all of the functionality is in [timeseries.py](https://github.com/datascopeanalytics/traces/blob/master/traces/timeseries.py)). It looks like pretty straight forward Python, but I have no real reference for how easily that would translate to go.

![cointop-bug](https://user-images.githubusercontent.com/69400937/132058730-193d841b-dd3a-4420-99fa-caf1e4a35890.png)

**Comment by blockjoe** [comment]
#159 

I threw out a small change here that takes care of tokens with missing data being added to the start and not the end, but that still operates under the assumption that there's alignment along the X axis of the data, and in the case of this current issue, would just move the drop off to the start and not the end.

This issue could be "fixed" by simply dropping the first point in the returned data slice if the API has returned some results with N values and other with N-1. That still has the underlying accuracy problem of the data potentially being unevenly sampled. I think for the accuracy that you get from an ASCII terminal plot, this shouldn't matter, but it does feel like a hacky fix to what might be better addressed by treating the data as a time series and not a list.

This "fix" could be as simple as

```go
var removeFirst bool
for i := range dataSlices {
    if len(dataSlices[i]) == baseLen - 1 {
        removeFirst = true
        break;
    }
}
if removeFirst {
    chart.SetData(data[1:])
} else {
    chart.SetData(data)
}
```

This has the problem of assuming that the issues with uneven API returns is strictly limited to the one off case and that the returns are aligned on the last point. Given that there is no meaningful x-axis to the chart, I think these tradeoffs would be unnoticeable.

**Comment by simon-anz** [comment]
Agree - data should be treated as time-series, with each coin being resampled to the same (consistent) timeframe, including zerp where the data is not present.  I have a separate Issue raised to add an X-axis, but that would require keeping the data (which is currently discarded before rendering)

**Comment by simon-anz** [comment]
@blockjoe traces looks like a good library, but I think it would difficult to port to golang - it seems to lack a lot of the language fundamentals that make the library work well on python.  I had a look for time-series stuff on golang, without much luck either. I don't really know golang!

**Comment by simon-anz** [comment]
Attempted to write a function to resample an array of [timestamp,value] using linear interpolation.

```
func interpolateTimeSeriesData(data [][]float64, start float64, end float64, numSteps int) [][]float64 {
	var newData [][]float64
	l := len(data)
	step := (end - start) / float64(numSteps)
	for pos := start; pos <= end; pos += step {
		idx := sort.Search(l, func(i int) bool { return data[i][0] >= pos })
		var val float64
		if idx == 0 {
			// off the left - use min value
			val = data[idx][1]
		} else if idx == l {
			// off the right - use max value
			val = data[idx-1][1]
		} else {
			// between two points - linear interpolation
			left := data[idx-1]
			right := data[idx]
			dv := right[1] - left[1]
			dt := right[0] - left[0]
			dvdt := float64(dv) / float64(dt)

			val = left[1] + (pos-left[0])*dvdt
		}
		newData = append(newData, []float64{pos, val})
	}
	return newData
}
```

This seems to work OK with quick test, but integrating it properly will require changing the way in which the data is cached (since it is currently stripped of timestamps first).

**Comment by simon-anz** [comment]
@blockjoe When the graphing is given coindata that doesn't include all of the data required - should it use 0 as value, or the oldest value it has?   There's a bit of an edge-case in that asking for 180D data will return a min-date that may be a bit after now-180D (so "missing" a sample or two?)

The code above uses the earliest value, which produces wrong-ish data.  With 10 steps on 180D char, the interpolation code above does:

```
CACHE1 HIT BTC
AFTER BTC
 2021-03-21 16:30:44 +1100 AEDT 57573.54
 2021-04-07 10:42:44 +1000 AEST 58003.67
 2021-04-24 05:54:44 +1000 AEST 51323.43
 2021-05-11 01:06:44 +1000 AEST 56758.37
 2021-05-27 20:18:44 +1000 AEST 39078.53
 2021-06-13 15:30:44 +1000 AEST 36463.40
 2021-06-30 10:42:44 +1000 AEST 35945.85
 2021-07-17 05:54:44 +1000 AEST 31455.36
 2021-08-03 01:06:44 +1000 AEST 39514.34
 2021-08-19 20:18:44 +1000 AEST 45757.63
 2021-09-05 15:30:44 +1000 AEST 49764.58
CACHE1 HIT GTC
AFTER GTC
 2021-03-21 16:30:44 +1100 AEDT 6.54
 2021-04-07 10:42:44 +1000 AEST 6.54
 2021-04-24 05:54:44 +1000 AEST 6.54
 2021-05-11 01:06:44 +1000 AEST 6.54
 2021-05-27 20:18:44 +1000 AEST 5.89
 2021-06-13 15:30:44 +1000 AEST 11.04
 2021-06-30 10:42:44 +1000 AEST 8.42
 2021-07-17 05:54:44 +1000 AEST 5.13
 2021-08-03 01:06:44 +1000 AEST 7.98
 2021-08-19 20:18:44 +1000 AEST 7.87
 2021-09-05 15:30:44 +1000 AEST 11.91
```

**Comment by simon-anz** [comment]
Here's my WIP code on this.  It caches the raw [timestamp,value] data from the backend, and resamples it on the fly. PR currently hardcodes "10" steps, and emits some logging as above.  

https://github.com/simon-anz/cointop/pull/1/files

The PR is actually based on the logrus branch #162 , since that makes logging infinitely easier ;)

**Comment by simon-anz** [comment]
OK, updated PR to return math.NaN() when resampling and the points fall outside the data range (which is about 1/3 of the data for GTC on 180D). Then remove the first+last point, since fetching "start-to-end" doesn't actually return start or end precisely (only within them).  Then replace NaN with 0 for the purpose of adding.

Still draft (too much debug output, only added to Portfolio charts, hardcode "10" resample points (should be maxX?) and based on logrus PR), but appreciate any input.

With a deliberately coarse 10, the portfolio for mostly GTC now looks like this - with correct?  almost-zero to the left.

![Screen Shot 2021-09-06 at 8 53 36 am](https://user-images.githubusercontent.com/69175652/132143550-93e1c1ca-7a74-4e02-a18a-ca809a3a6e91.png)

**Comment by simon-anz** [comment]
OK, I've cleaned up the code, and opened a PR. This code:
- collects the (raw) data from all portfolio entries before aggregating (this is where the cache is)
- examine the data to determine the time-interval between samples (say, 1D)
- calculates how many data points are needed for charting, then resample the data to this amount, using the time-range less one interval at each end

https://github.com/miguelmota/cointop/pull/165

![Screen Shot 2021-09-17 at 8 24 22 pm](https://user-images.githubusercontent.com/69175652/133768020-4b4c5d1d-9789-4c4a-b46c-4ecb65449176.png)

**Comment by lyricnz** [comment]
Solved by resample
